# Data Cleaning : Airbnb Listings
In this notebook we are going to deal with data from a well-known residential rental application, Airbnb. Specifically, the data for the Chicago area for 3 months of 2023 (January, February and March).

In [ ]:
#Harishobith

In [1]:
#Importing the necessary libraries
import numpy as np
import pandas as pd

In [2]:
# reading the csv files
listings_df = pd.read_csv("listings.csv")
reviews_df = pd.read_csv("reviews.csv")

## Data Cleaning : Listings.csv
1. We drop any entries that are missing (NaN) values.
2. We also drop any entries that are inconsistent such as price with a value of 0.

In [3]:
print(listings_df.shape)
listings_df.info()

(7747, 18)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7747 entries, 0 to 7746
Data columns (total 18 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   id                              7747 non-null   int64  
 1   name                            7747 non-null   object 
 2   host_id                         7747 non-null   int64  
 3   host_name                       7747 non-null   object 
 4   neighbourhood_group             0 non-null      float64
 5   neighbourhood                   7747 non-null   object 
 6   latitude                        7747 non-null   float64
 7   longitude                       7747 non-null   float64
 8   room_type                       7747 non-null   object 
 9   price                           7747 non-null   int64  
 10  minimum_nights                  7747 non-null   int64  
 11  number_of_reviews               7747 non-null   int64  
 12  last_review            

In [4]:
# determine the number of missing entries in each column of Listings.csv
for col in listings_df.columns:
    print (col + ', Number of Missing Values:', len(listings_df[col][listings_df[col].isnull()]))

id, Number of Missing Values: 0
name, Number of Missing Values: 0
host_id, Number of Missing Values: 0
host_name, Number of Missing Values: 0
neighbourhood_group, Number of Missing Values: 7747
neighbourhood, Number of Missing Values: 0
latitude, Number of Missing Values: 0
longitude, Number of Missing Values: 0
room_type, Number of Missing Values: 0
price, Number of Missing Values: 0
minimum_nights, Number of Missing Values: 0
number_of_reviews, Number of Missing Values: 0
last_review, Number of Missing Values: 1493
reviews_per_month, Number of Missing Values: 1493
calculated_host_listings_count, Number of Missing Values: 0
availability_365, Number of Missing Values: 0
number_of_reviews_ltm, Number of Missing Values: 0
license, Number of Missing Values: 1174


#### According to the above output, neighbourhood_group, last_review, reviews_per_month, and license columns have missing values.

In [5]:
# Checking whether this column has only null values
if listings_df['neighbourhood_group'].isna().all():
    print("neighbourhood_group has all NaN values")
else:
    print("neighbourhood_group does not have all NaN values")

neighbourhood_group has all NaN values


In [6]:
# As neighbourhood_group has only null values and cannot be replaced with any other values, so this column has to be dropped.
listings_df.drop('neighbourhood_group', axis = 1, inplace = True)
#listings_df.to_csv('listings.csv', index=False)

In [7]:
#Let's check whether there are any listings with price 0.00 (inconsistent values)
print ('Number of Listings with Price $0.00:', len(listings_df[listings_df['price'] == 0.00]))

Number of Listings with Price $0.00: 2


In [8]:
#As there are two listings with price 0.00, let's drop those two listings.
listings_df = listings_df[listings_df['price'] != 0.00]
print ('Number of Listings with Price $0.00:', len(listings_df[listings_df['price'] == 0.00]))

Number of Listings with Price $0.00: 0


In [9]:
# Then, let's drop all the records with null values as the missing values for last_review, reviews_per_month and license columns cannot be replaced with any other values.
listings_df = listings_df.dropna()

In [10]:
for col in listings_df.columns:
    print (col + ', Number of Missing Values:', len(listings_df[col][listings_df[col].isnull()]))

id, Number of Missing Values: 0
name, Number of Missing Values: 0
host_id, Number of Missing Values: 0
host_name, Number of Missing Values: 0
neighbourhood, Number of Missing Values: 0
latitude, Number of Missing Values: 0
longitude, Number of Missing Values: 0
room_type, Number of Missing Values: 0
price, Number of Missing Values: 0
minimum_nights, Number of Missing Values: 0
number_of_reviews, Number of Missing Values: 0
last_review, Number of Missing Values: 0
reviews_per_month, Number of Missing Values: 0
calculated_host_listings_count, Number of Missing Values: 0
availability_365, Number of Missing Values: 0
number_of_reviews_ltm, Number of Missing Values: 0
license, Number of Missing Values: 0


In [11]:
listings_df.shape

(5888, 17)

#### Our dataset is now clean and we no longer have any missing values or inconsistent values. We have a total of 5888 records from the original 7747; thus, we've retained a reasonable amount of records. 

## Data Cleaning : Reviews.csv

In [12]:
print(reviews_df.shape)
reviews_df.info()

(345939, 6)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 345939 entries, 0 to 345938
Data columns (total 6 columns):
 #   Column         Non-Null Count   Dtype 
---  ------         --------------   ----- 
 0   listing_id     345939 non-null  int64 
 1   id             345939 non-null  int64 
 2   date           345939 non-null  object
 3   reviewer_id    345939 non-null  int64 
 4   reviewer_name  345939 non-null  object
 5   comments       345847 non-null  object
dtypes: int64(3), object(3)
memory usage: 15.8+ MB


In [13]:
# determine the number of missing entries in each column of reviews.csv
for col in reviews_df.columns:
    print (col + ', Number of Missing Values:', len(reviews_df[col][reviews_df[col].isnull()]))

listing_id, Number of Missing Values: 0
id, Number of Missing Values: 0
date, Number of Missing Values: 0
reviewer_id, Number of Missing Values: 0
reviewer_name, Number of Missing Values: 0
comments, Number of Missing Values: 92


#### According to the above output, only comments column has 92 missing values and these missing values cannot be replaced by any other values. Hence, the rows with missing values has to be dropped.

In [14]:
reviews_df = reviews_df.dropna()
reviews_df.shape

(345847, 6)

In [15]:
for col in reviews_df.columns:
    print (col + ', Number of Missing Values:', len(reviews_df[col][reviews_df[col].isnull()]))

listing_id, Number of Missing Values: 0
id, Number of Missing Values: 0
date, Number of Missing Values: 0
reviewer_id, Number of Missing Values: 0
reviewer_name, Number of Missing Values: 0
comments, Number of Missing Values: 0


#### Reviews dataset doesn't have any missing values. Hence Reviews dataset has been cleaned successfully.

In [16]:
#Renaming the "id" column name of the listings dataframe so that it has the same name as the "listing_id" column of reviews dataframe.
listings_df = listings_df.rename(columns={'id': 'listing_id'})
listings_df.head()

listing_id                                            name  host_id  \
0        2384                    Hyde Park - Walk to UChicago     2613   
1       94450             Comfy Garden Suite in Andersonville   504470   
2      145659     Trendy Roscoe Village 3BR/2BR walk to shops   683529   
3        7126             Tiny Studio Apartment 94 Walk Score    17928   
4      189821  Best in Chicago, private, amazing garden space   899757   

  host_name neighbourhood  latitude  longitude        room_type  price  \
0   Rebecca     Hyde Park  41.78790  -87.58780     Private room     90   
1      Mark     Edgewater  41.97960  -87.66512  Entire home/apt     65   
2       Joe  North Center  41.94342  -87.68121  Entire home/apt    198   
3     Sarah     West Town  41.90166  -87.68021  Entire home/apt     85   
4   Meighan  Logan Square  41.92918  -87.70219  Entire home/apt    202   

   minimum_nights  number_of_reviews last_review  reviews_per_month  \
0               3                212  2023-03-04               2.13   
1               7                  7  2022-12-01               0.88   
2               2                 59  2022-12-31               0.65   
3               2                483  2023-02-26               2.89   
4               3                598  2023-02-25               4.27   

   calculated_host_listings_count  availability_365  number_of_reviews_ltm  \
0                               1               347                     20   
1                               1               178                      7   
2                              22               330                     15   
3                               1               279                     48   
4                               1               207                     24   

        license  
0  R17000015609  
1  R22000082956  
2       2209272  
3  R21000075737  
4  R21000062936

### Merging the two dataframes

In [19]:
# Merging the two dataframes with the common column "listing_id".
merged_df = pd.merge(listings_df, reviews_df, on='listing_id')
merged_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 328458 entries, 0 to 328457
Data columns (total 22 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   listing_id                      328458 non-null  int64  
 1   name                            328458 non-null  object 
 2   host_id                         328458 non-null  int64  
 3   host_name                       328458 non-null  object 
 4   neighbourhood                   328458 non-null  object 
 5   latitude                        328458 non-null  float64
 6   longitude                       328458 non-null  float64
 7   room_type                       328458 non-null  object 
 8   price                           328458 non-null  int64  
 9   minimum_nights                  328458 non-null  int64  
 10  number_of_reviews               328458 non-null  int64  
 11  last_review                     328458 non-null  object 
 12  reviews_per_mont